### PR TITLE
[PM-4614] Updating Duo to SDK v4 for Universal Prompt

### PR DESCRIPTION
**Pull request by ike-kottlowski**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Create a service and feature flag to support the DUO v4 SDK Universal Prompt.

## Code changes

* **src/Core/Auth/Identity/TemporaryDuoWebV4SDKService.cs:** Added new service to fill the gaps to handle the DUO v4 SDK update.
* **src/Identity/IdentityServer/BaseRequestValidator.cs:** Updated DUO two factor flows to support DUO v4 SDK.
  * On creation we check for the feature flag `duo-redirect` which adds the AuthUrl to the return object
  * On Validation we check the token for a `:` which is indicative of DUO SDK v2. If `:` in not present then we validate against the new `TemporaryDuoWebV4SDKService`
* **src/Core/Constants.cs:** Added `duo-redirect` feature flag

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Identity/IdentityServer/BaseRequestValidator.cs [hunk 1]
```diff
@@ -465,11 +465,25 @@ private Device GetDeviceFromRequest(ValidatedRequest request)
...
                     {
-                        ["Host"] = provider.MetaData["Host"],
-                        ["Signature"] = token
-                    };
+                        CurrentContext.HttpContext.Request.Headers.TryGetValue("X-Bitwarden-Client-Name", out var bitwardenClientName);
+                        return new Dictionary<string, object>
+                        {
+                            ["Host"] = provider.MetaData["Host"],
+                            ["Duo2faInitiatingClient"] = bitwardenClientName,
+                            ["AuthUrl"] = token
```
Comment (jlf0dev) [inline src/Identity/IdentityServer/BaseRequestValidator.cs:476]: Can we append the client name to the auth url before sending down to client?

Comment (ike-kottlowski) [reply src/Identity/IdentityServer/BaseRequestValidator.cs:476]: I assumed we could use the `Dui2faInitiatingClient` to more directly track the client name instead of having to parse the AuthUrl.

Comment (ike-kottlowski) [reply src/Identity/IdentityServer/BaseRequestValidator.cs]: appended to the redirect.

File: src/Core/Auth/Identity/DuoUniversalPromptService.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    /// <summary>
+    /// Provider agnostic (either Duo or OrganizationDuo) method to generate a Duo Auth URL
+    /// </summary>
+    /// <param name="provider">Either Duo or OrganizationDuo</param>
+    /// <param name="user">self</param>
+    /// <returns>AuthUrl for DUO SDK v4</returns>
+    public async Task<string> GenerateAsync(TwoFactorProvider provider, User user)
+    {
+        if (!HasProperMetaData(provider)) return null;
```
Comment (jlf0dev) [inline src/Core/Auth/Identity/DuoUniversalPromptService.cs:47]: Our code style [requires braces for conditionals](https://contributing.bitwarden.com/contributing/code-style/csharp#control-blocks), there's a couple instances of this throughout the PR.

File: src/Core/Auth/Identity/DuoUniversalPromptService.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        return provider?.MetaData != null && provider.MetaData.ContainsKey("IKey") &&
+            provider.MetaData.ContainsKey("SKey") && provider.MetaData.ContainsKey("Host");
+    }
+
+    /// <summary>
+    /// Generates a Duo.Client object for use with Duo SDK v4. This combines the health check and the client generation
+    /// </summary>
+    /// <param name="provider">TwoFactorProvider Duo or OrganizationDuo</param>
+    /// <returns></returns>
```
Comment (jlf0dev) [inline src/Core/Auth/Identity/DuoUniversalPromptService.cs:86]: empty `<returns></returns>`

File: src/Core/Auth/Identity/OrganizationDuoWebTokenProvider.cs
Comment (jlf0dev) [inline src/Core/Auth/Identity/OrganizationDuoWebTokenProvider.cs:1]: We should likely revert these changes in order to avoid introducing bugs outside of the feature flag

File: src/Core/Auth/Identity/DuoWebTokenProvider.cs
Comment (jlf0dev) [inline src/Core/Auth/Identity/DuoWebTokenProvider.cs:1]: Should probably revert these changes as well

File: src/Identity/IdentityServer/BaseRequestValidator.cs [hunk 4]
```diff
@@ -428,10 +431,23 @@ private Device GetDeviceFromRequest(ValidatedRequest request)
...
                     return false;
                 }
+                // DUO SDK v4 Update: try to validate the token - PM-5156 addresses tech debt
+                if (FeatureService.IsEnabled(FeatureFlagKeys.DuoRedirect, CurrentContext))
+                {
+                    if (type == TwoFactorProviderType.Duo)
+                    {
+                        if (!token.Contains(':'))
+                        {
+                            // We have to send the provider to the DuoUniversalPromptService to get create the DuoClient
```
Comment (jlf0dev) [inline src/Identity/IdentityServer/BaseRequestValidator.cs:445]:
```suggestion
                            // We have to send the provider to the DuoUniversalPromptService to create the DuoClient
```

File: src/Identity/IdentityServer/BaseRequestValidator.cs [hunk 5]
```diff
@@ -465,11 +495,19 @@ private Device GetDeviceFromRequest(ValidatedRequest request)
...
+                    var duoResponse = new Dictionary<string, object>
                     {
                         ["Host"] = provider.MetaData["Host"],
                         ["Signature"] = token
                     };
+
+                    // DUO SDK v4 Update: Duo-Redirect
+                    if (FeatureService.IsEnabled(FeatureFlagKeys.DuoRedirect, CurrentContext))
+                    {
+                        // Use generate token from new token provider
```
Comment (jlf0dev) [inline src/Identity/IdentityServer/BaseRequestValidator.cs:507]:
```suggestion
                        // Generate auth url from new token provider
```
Hopefully adds some value to the comment, ignore if its not what you meant

File: src/Core/Auth/Identity/DuoUniversalPromptService.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+namespace Bit.Core.Auth.Identity;
+
+/* 
+    PM-5156 addresses tech debt
+    Interface to allow for DI, will end up being removed as part of the removal of the old Duo SDK v2 flows.
+    This service is to support SDK v4 flows for Duo. At some time in the future we will need
+    to combine this service with the DuoWebTokenProvider and OrganizationDuoWebTokenProvider to support SDK v4.
+*/
+public interface IDuoUniversalPromptService
```
Comment (jlf0dev) [inline src/Core/Auth/Identity/DuoUniversalPromptService.cs:15]: Can we add temporary to the name here? Something like `ITemporaryDuoSDKUpgradeService`? The universal prompt is actually separate from our work here. It's more of a upgrade from an iframe implementation to a frameless implementation.

**Review by jlf0dev — CHANGES_REQUESTED** [review]

File: src/Identity/IdentityServer/BaseRequestValidator.cs [hunk 7]
```diff
@@ -428,10 +431,23 @@ private Device GetDeviceFromRequest(ValidatedRequest request)
...
                     return false;
                 }
+                // DUO SDK v4 Update: try to validate the token - PM-5156 addresses tech debt
+                if (FeatureService.IsEnabled(FeatureFlagKeys.DuoRedirect))
+                {
+                    if (type == TwoFactorProviderType.Duo)
+                    {
+                        if (!token.Contains(':'))
+                        {
+                            // We have to send the provider to the DuoUniversalPromptService to create the DuoClient
```
Comment (jlf0dev) [inline src/Identity/IdentityServer/BaseRequestValidator.cs:445]: nit: missed a rename

**Review by jlf0dev — APPROVED** [review]

**Review by jlf0dev — APPROVED** [review]
